File: mobile/modules/share-to-app/ios/ShareToAppModule.swift
import ExpoModulesCore
import OSLog
import SCSDKCreativeKit
import UIKit

/**
 Why a share fell back, in Console.

 Every failure path here resolves `false` so the caller can drop to the share
 sheet, which is the right behaviour and also completely silent — the user sees
 the system sheet and cannot tell whether Snapchat is missing, the client ID is
 unapproved, or the card failed to render. All three look identical, and the
 last two are our bugs. Logged at `error` level so it survives in a release
 build and can be read with `log stream --predicate 'subsystem == "app.nutriai.mobile"'`.
 */
private let shareLog = Logger(subsystem: "app.nutriai.mobile", category: "share-to-app")

/**
 Also NSLog, not only os_log.

 os_log is the right home for this — it survives in release builds and is
 readable in Console.app — but it is *not* forwarded to a device's stdout, which
 is the only stream reachable over a cable with `devicectl process launch
 --console`. Diagnosing this on a real phone meant seeing nothing at all. NSLog
 reaches both, and the volume here is a handful of lines per share.
 */
private func shareDiag(_ message: String) {
  shareLog.error("\(message, privacy: .public)")
  NSLog("[share-to-app] %@", message)
}

/**
 Whether Snapchat is even on this phone.

 Creative Kit fails with the same opaque error whether Snapchat is missing or
 the client ID is unapproved, and those are wildly different problems: one is a
 config fix, the other means there is nothing to debug. `canOpenURL` needs
 `snapchat` in LSApplicationQueriesSchemes, which withSnapCreativeKit adds.
 */

/**
 Turns whatever the JS side hands us into a readable file URL.

 react-native-view-shot returns a *bare filesystem path* on iOS
 (`/private/var/mobile/.../tmp/ReactNative/xxx.png`), while Android returns a
 `content://` URI. `URL(string:)` requires a scheme, so it quietly produced a
 useless relative URL for the iOS form, `Data(contentsOf:)` failed, and the
 share bailed out *before Creative Kit was ever called* — which looked
 identical to Snapchat rejecting us, and sent the debugging off after the
 client ID and the portal for hours.

 `URL(fileURLWithPath:)` is the right constructor for a path, and the scheme
 check is what decides which one applies.
 */
private func readableFileURL(_ uri: String) -> URL? {
  if let url = URL(string: uri), url.scheme != nil { return url }
  return URL(fileURLWithPath: uri)
}

/**
 Keeps the Creative Kit API alive until it answers.

 `SCSDKSnapAPI` does not retain itself across `startSending`. Created as a local
 it is released the moment the enclosing block returns, and the completion
 handler is then simply never called — no error, no callback, nothing. On screen
 that was a share button stuck on "Preparing…" forever, and in the log it was a
 line saying the send had started followed by silence.

 Held here for the lifetime of the call and cleared in the completion. Only one
 share can be in flight at a time (the UI disables itself while `sharing` is
 true), so a single slot is enough, and it is only ever touched on the main
 queue.
 */
private var inFlightSnapAPI: SCSDKSnapAPI?

private func snapchatInstalled() -> Bool {
  guard let url = URL(string: "snapchat://") else { return false }
  return UIApplication.shared.canOpenURL(url)
}

/**
 Hands a captured card to Snapchat as a *Snap*, not a chat attachment.

 The iOS half of the Android module next door, and deliberately not a port of
 it: Android can reach Snapchat's camera preview with a bare Intent, while iOS
 has no equivalent — the sandbox offers no way to hand content to a named app,
 so Creative Kit's SDK is the only route. What the two share is the outcome and
 the failure mode.

 The failure mode is worth stating, because it is invisible: `UIActivityViewController`
 (the system share sheet) *does* offer Snapchat, and Snapchat *does* accept the
 image — as a chat message, with no editor and no Story option. It looks like a
 working share right up until someone receives one. That is what this replaces.

 Everything here degrades rather than throws. The JS caller branches on the
 returned boolean and falls back to the share sheet, so returning false is how
 this says "use the old path" — a thrown error would break that chain and leave
 the user with a button that does nothing.
 */
public class ShareToAppModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ShareToApp")

    /**
     Android reports which apps are installed so the caller can hide a dead
     button. iOS cannot answer that honestly without declaring every scheme in
     `LSApplicationQueriesSchemes`, and the answer would not change what this
     module does — Creative Kit reports its own failure when Snapchat is
     missing. Always false, matching the JS contract.
     */
    Function("isAppInstalled") { (_: String) -> Bool in
      false
    }

    /**
     The Android-only direct-send intent. No iOS equivalent exists; the JS side
     never calls this on iOS, and it is defined only so the native module's
     shape is identical on both platforms.
     */
    AsyncFunction("shareImage") { (_: String, _: String, _: String) -> Bool in
      false
    }

    AsyncFunction("shareSnapToPreview") { (
      uri: String,
      clientId: String,
      appName: String,
      caption: String?,
      promise: Promise
    ) in
      guard !clientId.isEmpty else {
        shareDiag("snap preview: no Creative Kit client ID in this build")
        promise.resolve(false)
        return
      }

      /**
       The card is captured to a file by react-native-view-shot, so it arrives
       as a URL rather than an image. Loading it here rather than passing the
       URL to `SCSDKSnapPhoto(imageUrl:)` is deliberate: that initialiser
       expects a URL the SDK can read later, and a temp file the OS may have
       already reclaimed produces an empty preview — the same "card failed to
       render" symptom the Android side has to guard against with an explicit
       URI permission grant.
       */
      guard
        let url = readableFileURL(uri),
        let data = try? Data(contentsOf: url),
        let image = UIImage(data: data)
      else {
        shareDiag("snap preview: could not load the captured card from \(uri)")
        promise.resolve(false)
        return
      }

      DispatchQueue.main.async {
        shareDiag("snap preview: clientId=\(clientId.prefix(8))… snapchatInstalled=\(snapchatInstalled())")
        let photo = SCSDKSnapPhoto(image: image)
        let content = SCSDKPhotoSnapContent(snapPhoto: photo)

        if let caption, !caption.isEmpty {
          content.caption = caption
        }
        /**
         The swipe-up link on the Snap.

         This is the native version of the download URL that used to be printed
         along the bottom of the card: a recipient swipes up and lands on the
         install page, rather than squinting at a URL and typing it. It also
         means the card art no longer has to carry it.
         */
        content.attachmentUrl = "https://nutriai-app.up.railway.app/download"

        let api = SCSDKSnapAPI()
        inFlightSnapAPI = api
        api.startSending(content) { error in
          inFlightSnapAPI = nil
          // Snapchat missing, too old, or a client ID the portal has not
          // approved for this bundle. All of them mean "fall back" — but they
          // are very different bugs, so say which one.
          if let error {
            shareDiag("snap preview FAILED: \(String(describing: error))")
          } else {
            shareDiag("snap preview opened OK")
          }
          promise.resolve(error == nil)
        }
      }
    }

    /**
     Opens Snapchat's camera with the card riding as a sticker on top.

     `SCSDKNoSnapContent` is the iOS equivalent of Android's camera deep link:
     content with no photo of its own, which lands the user in the camera rather
     than the preview. The sticker is the payload, and the background is
     whatever they shoot or pick.

     `width` and `height` are in *points* and must be set explicitly. Left
     unset, Creative Kit sizes the sticker from the UIImage, which is in
     *pixels* — a 1080×1920 export became a 1080×1920-point sticker on a
     393-point-wide screen, so only a fragment of it was visible and the rest
     ran off every edge. The caller knows the frame it wants to fill, so the
     size comes from there rather than from the bitmap.
     */
    AsyncFunction("shareSnapSticker") { (
      uri: String,
      clientId: String,
      appName: String,
      widthDp: Int,
      heightDp: Int,
      posY: Double,
      promise: Promise
    ) in
      guard !clientId.isEmpty else {
        shareDiag("snap sticker: no Creative Kit client ID in this build")
        promise.resolve(false)
        return
      }

      guard
        let url = readableFileURL(uri),
        let data = try? Data(contentsOf: url),
        let image = UIImage(data: data)
      else {
        shareDiag("snap sticker: could not load the captured sticker from \(uri)")
        promise.resolve(false)
        return
      }

      DispatchQueue.main.async {
        let content = SCSDKNoSnapContent()
        let sticker = SCSDKSnapSticker(stickerImage: image)
        sticker.width = CGFloat(widthDp)
        sticker.height = CGFloat(heightDp)
        sticker.posX = 0.5
        sticker.posY = CGFloat(posY)
        content.sticker = sticker
        content.attachmentUrl = "https://nutriai-app.up.railway.app/download"

        let api = SCSDKSnapAPI()
        inFlightSnapAPI = api
        api.startSending(content) { error in
          inFlightSnapAPI = nil
          if let error {
            shareDiag("snap sticker FAILED: \(String(describing: error))")
          } else {
            shareDiag("snap sticker opened OK")
          }
          promise.resolve(error == nil)
        }
      }
    }
  }
}

File: mobile/modules/share-to-app/android/src/main/java/app/nutriai/sharetoapp/ShareToAppModule.kt
package app.nutriai.sharetoapp

import android.app.PendingIntent
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import org.json.JSONObject
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Hands an image to a *named* app via ACTION_SEND.
 *
 * This exists because expo-intent-launcher cannot express the intent Android
 * actually requires. ACTION_SEND carries its payload in `EXTRA_STREAM`, which
 * must be a Parcelable `Uri`; the JS bridge can only put primitives in extras,
 * so the Uri arrived as a String and the intent failed every time — verified
 * on a device, where the "direct" share silently fell through to the system
 * chooser. Building the Intent in Kotlin is the only way to put a real Uri in
 * there.
 *
 * Android-only by design. iOS has no equivalent: handing content to a specific
 * app needs that app's SDK (Snap's Creative Kit), and the sandbox offers no
 * general route, so the JS side falls back to the share sheet there.
 */
class ShareToAppModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ShareToApp")

    /**
     * True when the package is installed *and* visible to us. On Android 11+
     * visibility requires a <queries> entry in the manifest — without it this
     * returns false even for an installed app, which is exactly the failure
     * that makes a share button look broken rather than absent.
     */
    Function("isAppInstalled") { packageName: String ->
      val pm = appContext.reactContext?.packageManager ?: return@Function false
      try {
        pm.getPackageInfo(packageName, 0)
        true
      } catch (e: PackageManager.NameNotFoundException) {
        false
      }
    }

    /**
     * Throws rather than returning false on failure: the JS caller falls back
     * to the system share sheet, and a silent no-op would leave the user
     * tapping a button that does nothing.
     */
    AsyncFunction("shareImage") { uri: String, packageName: String, mimeType: String ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("No application context")

      val intent = Intent(Intent.ACTION_SEND).apply {
        type = mimeType
        setPackage(packageName)
        // The whole point of this module: a real Uri, not its toString().
        putExtra(Intent.EXTRA_STREAM, Uri.parse(uri))
        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        // Started from outside an Activity context.
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      }

      // resolveActivity respects package visibility, so this also catches the
      // missing-<queries> case with a message that says what is wrong.
      if (intent.resolveActivity(context.packageManager) == null) {
        throw IllegalStateException(
          "$packageName cannot receive $mimeType (not installed, or missing a <queries> entry)"
        )
      }

      context.startActivity(intent)
      true
    }

    /**
     * Sends the card to Snapchat as a *Snap*, not a chat attachment.
     *
     * `shareImage` above hands Snapchat a plain ACTION_SEND, which its generic
     * share receiver turns into the "Send To" flow — the picture arrives in a
     * chat, as a message. That is not what a share button on a story card is
     * for, and it was the behaviour on a real device.
     *
     * Creative Kit Lite is the documented route to the camera preview: the
     * editor where the recipient list, the Story option and Snapchat's own
     * creative tools live. Mechanically it is the same ACTION_SEND, and the
     * three additions below are the entire difference:
     *
     *   - the intent *data* is `snapchat://creativekit/preview`, which selects
     *     the preview entry point rather than the share receiver;
     *   - `CLIENT_ID` identifies the calling app. Snapchat ignores the deep
     *     link without it and falls back to the chat flow, which is why this
     *     cannot be done anonymously;
     *   - `CLIENT_APP_NAME` is the attribution Snapchat prints on the Snap.
     *
     * See https://github.com/Snapchat/creative-kit (Creative Kit Lite). No SDK
     * is involved — the contract is this Intent.
     */
    AsyncFunction("shareSnapToPreview") { uri: String, clientId: String, appName: String, caption: String? ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("No application context")

      if (clientId.isBlank()) {
        throw IllegalStateException("No Snap Creative Kit client ID configured")
      }

      val fileUri = Uri.parse(uri)
      val intent = Intent(Intent.ACTION_SEND).apply {
        setPackage(SNAPCHAT_PACKAGE)
        // setDataAndType, not `type =`: setting the type alone clears the data
        // Uri, and losing it is exactly what silently demotes this back to a
        // chat attachment.
        setDataAndType(Uri.parse(CREATIVE_KIT_PREVIEW), "image/*")
        putExtra(CLIENT_ID_EXTRA, clientId)
        putExtra(CLIENT_APP_NAME_EXTRA, appName)
        putExtra(Intent.EXTRA_STREAM, fileUri)
        if (!caption.isNullOrBlank()) putExtra(CAPTION_TEXT_EXTRA, caption)
        /**
         * How Snapchat gets back to us when the user finishes or cancels.
         *
         * Snap's reference implementation sets this on every Creative Kit
         * intent, so it is part of the contract rather than an optional extra —
         * matched here rather than trimmed, because the failure mode for
         * getting this wrong is Snapchat quietly treating the intent as an
         * ordinary share, which is indistinguishable from the bug being fixed.
         *
         * FLAG_IMMUTABLE is required from Android 12; the PendingIntent carries
         * an empty Intent because we want the callback, not a destination.
         */
        putExtra(
          RESULT_INTENT_EXTRA,
          PendingIntent.getActivity(
            context,
            CREATIVE_KIT_REQUEST_CODE,
            Intent(),
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) PendingIntent.FLAG_IMMUTABLE
            else PendingIntent.FLAG_ONE_SHOT
          )
        )
        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
      }

      if (intent.resolveActivity(context.packageManager) == null) {
        throw IllegalStateException(
          "Snapchat cannot receive a Creative Kit preview (not installed, too old, or missing a <queries> entry)"
        )
      }

      /**
       * The flag alone is not always enough.
       *
       * FLAG_GRANT_READ_URI_PERMISSION grants against the intent's *data* Uri,
       * and here the data is the creativekit deep link — the image rides in
       * EXTRA_STREAM instead. An explicit grant to Snapchat covers it; without
       * this the preview can open on an empty canvas, which looks like the card
       * failed to render rather than like a permissions problem.
       */
      context.grantUriPermission(SNAPCHAT_PACKAGE, fileUri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
      context.startActivity(intent)
      true
    }

    /**
     * Opens Snapchat's *camera* with the card attached as a sticker.
     *
     * The other direction entirely from `shareSnapToPreview`. There, our card
     * is the Snap and the user's only choice is whether to send it. Here the
     * card is an overlay and the background is theirs — they point the camera
     * at the meal, or pick any photo, and our design floats on top of it.
     *
     * Worth the second entry point because it changes what the thing *is*: a
     * full-frame card posted to a Story is an advert with someone's numbers on
     * it, while a sticker over their own photo is their post, which we happened
     * to caption. The second gets shared; the first gets skipped.
     *
     * The sticker PNG must have a transparent background — see the note in
     * StickerFrame.tsx. A card captured with its own background works here but
     * looks like a screenshot pasted over the photo.
     *
     * Position and size are fractions and dp respectively, matching Snap's
     * documented sticker JSON. The defaults put it slightly above centre, where
     * it covers the least of a typical hand-held food shot.
     */
    AsyncFunction("shareSnapSticker") {
      uri: String,
      clientId: String,
      appName: String,
      widthDp: Int,
      heightDp: Int,
      posY: Double
      ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("No application context")

      if (clientId.isBlank()) {
        throw IllegalStateException("No Snap Creative Kit client ID configured")
      }

      val stickerUri = Uri.parse(uri)
      val sticker = JSONObject().apply {
        put("uri", stickerUri.toString())
        put("posX", 0.5)
        put("posY", posY)
        put("rotation", 0)
        put("widthDp", widthDp)
        put("heightDp", heightDp)
      }

      val intent = Intent(Intent.ACTION_SEND).apply {
        setPackage(SNAPCHAT_PACKAGE)
        // The camera deep link, not the preview one: that is the whole
        // difference between "here is your Snap" and "here is your sticker,
        // now go take a Snap".
        setDataAndType(Uri.parse(CREATIVE_KIT_CAMERA), INTENT_TYPE_ALL)
        putExtra(CLIENT_ID_EXTRA, clientId)
        putExtra(CLIENT_APP_NAME_EXTRA, appName)
        putExtra(STICKER_EXTRA, sticker.toString())
        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        putExtra(
          RESULT_INTENT_EXTRA,
          PendingIntent.getActivity(
            context,
            CREATIVE_KIT_REQUEST_CODE,
            Intent(),
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) PendingIntent.FLAG_IMMUTABLE
            else PendingIntent.FLAG_ONE_SHOT
          )
        )
      }

      if (intent.resolveActivity(context.packageManager) == null) {
        throw IllegalStateException(
          "Snapchat cannot receive a Creative Kit sticker (not installed, too old, or missing a <queries> entry)"
        )
      }

      // The sticker rides in an extra rather than the intent data, so the flag
      // above does not cover it. Same reasoning as the preview flow.
      context.grantUriPermission(SNAPCHAT_PACKAGE, stickerUri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
      context.startActivity(intent)
      true
    }
  }

  private companion object {
    const val SNAPCHAT_PACKAGE = "com.snapchat.android"
    const val CREATIVE_KIT_PREVIEW = "snapchat://creativekit/preview"
    const val CREATIVE_KIT_CAMERA = "snapchat://creativekit/camera"
    const val INTENT_TYPE_ALL = "*/*"
    const val STICKER_EXTRA = "sticker"
    const val CLIENT_ID_EXTRA = "CLIENT_ID"
    const val CLIENT_APP_NAME_EXTRA = "CLIENT_APP_NAME"
    const val CAPTION_TEXT_EXTRA = "captionText"
    const val RESULT_INTENT_EXTRA = "RESULT_INTENT"
    const val CREATIVE_KIT_REQUEST_CODE = 100
  }
}
